receiver/loadgenreceiver/internal/list/bounded_looping.go

```diff
@@ -0,0 +1,52 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package list // import "github.com/elastic/opentelemetry-collector-components/receiver/loadgenreceiver/internal/list"
+
+import (
+	"errors"
+	"sync"
+)
+
+var ErrLoopLimitReached = errors.New("loop limit reached")
+
+// BoundedLoopingList is a list that can optionally have a limit on the number of loops performed.
+type BoundedLoopingList[T any] struct {
+	list      LoopingList[T]
+	loopLimit int
+	mu        sync.Mutex
+}
+
+func NewBoundedLoopingList[T any](items []T, loopLimit int) BoundedLoopingList[T] {
+	return BoundedLoopingList[T]{
+		list:      LoopingList[T]{items: items},
+		loopLimit: loopLimit,
+	}
+}
+
+// Next returns the next element in the looping list.
+// When loop limit is reached, it returns a zero value of T and ErrLoopLimitReached.
+// Safe for concurrent use.
+func (s *BoundedLoopingList[T]) Next() (T, error) {
```

Comment:
Why is the BoundedLoopingList made safe for concurrent use vs. the LoopingList isn't? Logically from the perspective of a user the difference between them is that one loops infinitely, whereas the other one doesn't - it's not clear to me why this would imply difference in safety for concurrent usage.

Reply:
Given the two are pretty much the same and it is not worth the extra complexity to separate them, I've combined them to reduce confusion.